docs/DebuggingTheCompiler.md

```diff
@@ -553,6 +553,95 @@ the pipeline. Here is a quick summary of the various options:
   When printing IR for functions for print-[before|after]-all options, Only
   print the IR for functions whose name is in this comma separated list.
 
+## Debugging assembly and object code
+
+Understanding layout of compiler-generated metadata
+can sometimes involve looking at assembly and object code.
+
+### Working with a single file
+
+Here's how to generate assembly or object code:
+
+```
+# Emit assembly in Intel syntax (AT&T syntax is the default)
+swiftc tmp.swift -emit-assembly -Xllvm -x86-asm-syntax=intel -o tmp.S
+
+# Emit object code
+swiftc tmp.swift -emit-object -o tmp.o
+```
+
+Understanding mangled names can be hard though: `swift demangle` to the rescue!
+
+```
+swiftc tmp.swift -emit-assembly -Xllvm -x86-asm-syntax=intel -o - \
+  | swift demangle > tmp-demangled.S
+
+swiftc tmp.swift -emit-object -o tmp.o
+
+# Look at where different symbols are located, sorting by address (-n)
+# and displaying section names (-m)
+nm -n -m tmp.o | swift demangle > tmp.txt
+
+# Inspect disassembly of an existing dylib (AT&T syntax is the default)
+objdump -d --x86-asm-syntax=intel /path/to/libcake.dylib \
+  | swift demangle > libcake.S
+```
+
+### Working with multiple files
+
+Some bugs only manifest in WMO, and may involve complicated Xcode projects.
+Moreover, Xcode may be passing arguments via `-filelist`
+and expecting outputs via `-output-filelist`, and those file lists
+may be in temporary directories.
+
+If you want to inspect assembly or object code for individual files when
+compiling under WMO, you can mimic this by doing the following:
```

Comment:
Right, by default WMO will use a single object file. Xcode is ensuring that N object files are created by passing in a `-output-file-map /path/toProjectName-OutputFileMap.json` which contains paths for the various files such as diagnostic files, dependency files and object files. Since the JSON format is a bit more work to create, I thought it would be better to stick to the simpler text based format. However, `-output-filelist` is a frontend flag, which is why I suggest invoking the frontend directly here.

Reply:
In the past I've created a tool to generate an output file map for some given source files, and the output paths used convention. Being a tool it also could create intermediate directories. I think it could be reasonable to include such a tool, or even put some simple-case logic into the driver. But at this point in time, I guess the frontend and these files are the way to go.